File: backend/src/models/service.ts
import { DocumentReference } from "firebase-admin/firestore";
import { firestore } from "../config/firebase.config";
import {
  DynamicObject,
  classToApi,
  dateFromAlgolia,
  withoutNulls,
} from "./utilities/util";
import Media from "./utilities/file";

export default class Service {
  serviceId?: string;
  name?: string;
  description?: string;
  topic?: string;
  image?: string; // This could be a URL string pointing to the image location
  subServices?: DocumentReference[]; // List of sub-services which are also of type Service
  parentServiceId?: string | null; // This could be null if the service is a root service
  isRoot?: boolean; // Flag to indicate if this is a root service
  createdAt?: Date; // Flag to indicate if this is a root service
  createdBy?: DocumentReference;
  documentRef?: DocumentReference;
  constructor({
    serviceId,
    name,
    description,
    topic,
    image,
    createdAt,
    subServices = [],
    parentServiceId = null,
    isRoot = false,
    createdBy,
    documentRef,
  }: {
    serviceId?: string;
    name?: string;
    description?: string;
    topic?: string;
    image?: string;
    createdAt?: Date;
    subServices?: DocumentReference[];
    createdBy?: DocumentReference;
    documentRef?: DocumentReference;
    parentServiceId?: string | null;
    isRoot?: boolean;
  }) {
    this.serviceId = serviceId;
    this.name = name;
    this.description = description;
    this.topic = topic;
    this.createdAt = createdAt;
    this.image = image;
    this.subServices = subServices;
    this.parentServiceId = parentServiceId;
    this.isRoot = isRoot;
    this.createdBy = createdBy;
    this.documentRef = documentRef;
  }

  toFirestoreObject(): firestore.DocumentData {
    return withoutNulls({
      serviceId: this.serviceId,
      name: this.name,
      description: this.description,
      topic: this.topic,
      image: this.image,
      created_at: this.createdAt,
      subServices: this.subServices,
      parentServiceId: this.parentServiceId,
      createdBy: this.createdBy,
      isRoot: this.isRoot,
    });
  }
  static fromAlgolia(data: DynamicObject): Service {
    return new Service({
      serviceId: data.serviceId,
      name: data.name,
      description: data.description,
      topic: data.topic,
      image: data.image,
      subServices: data.subServices,
      parentServiceId: data.parentServiceId,
      isRoot: data.isRoot,
      createdBy: data.createdBy,
      createdAt: dateFromAlgolia(data.created_at),
      documentRef: data.path.match(/\/(.*)$/)[1],
    });
  }

  static async ServiceToApi(s: Service) {
    return await classToApi(s);
  }
  // Methods to manage service data could be added here, or in a separate service class
}

File: mobile/lib/data/models/service/service.dart
class Service {
  String? serviceId;
  String? name;
  String? description;
  String? topic;
  String? image;
  List<String>? subServices;
  String? parentServiceId;
  bool? isRoot;
  DateTime? createdAt;
  String? createdBy;
  String? documentRef;

  Service({
    this.serviceId,
    this.name,
    this.description,
    this.topic,
    this.image,
    this.subServices,
    this.parentServiceId,
    this.isRoot,
    this.createdAt,
    this.createdBy,
    this.documentRef,
  });

  // Convert JSON to Service object
  factory Service.fromJson(Map<String, dynamic> json) => Service(
        serviceId: json['serviceId'],
        name: json['name'],
        description: json['description'],
        topic: json['topic'],
        image: json['image'],
        subServices: json['subServices'] != null
            ? List<String>.from(json['subServices'])
            : null,
        parentServiceId: json['parentServiceId'],
        isRoot: json['isRoot'],
        createdAt: json['createdAt'] != null
            ? DateTime.parse(json['createdAt'])
            : null,
        createdBy: json['createdBy'],
        documentRef: json['documentRef'],
      );

  // Convert Service object to JSON
  Map<String, dynamic> toJson() => {
        'serviceId': serviceId,
        'name': name,
        'description': description,
        'topic': topic,
        'image': image,
        'subServices': subServices,
        'parentServiceId': parentServiceId,
        'isRoot': isRoot,
        'createdAt': createdAt != null ? createdAt!.toIso8601String() : null,
        'createdBy': createdBy,
        'documentRef': documentRef,
      };
}
